extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java

```diff
@@ -0,0 +1,204 @@
+     * the incoming URI..
+     */
+    public static final String DEFAULT_URI_PROTOCOL = "ssh";
+
+    /**
+     * The default host to use if one is not defined.
+     */
+    public static final String DEFAULT_URI_HOST = "localhost";
+
+    /**
+     * The regex to use to split username and password.
+     */
+    private static final Pattern userinfoPattern = Pattern.compile("(^[^:]+):?(.*)");
+
+    /**
+     * The regex group of the username.
+     */
+    private static final int USERNAME_GROUP = 1;
+
+    /**
+     * THe regex group of the password.
+     */
+    private static final int PASSWORD_GROUP = 2;
+
+    /**
+     * Parse out a URI string and get a connection from that string,
+     * or an exception if the parsing fails.
+     *
+     * @param uri
+     *     The string form of the URI to be parsed.
+     *
+     * @return
+     *     A GuacamoleConfiguration using a combination of the parsed
+     *     URI values and default values when not specified in the
+     *     URI.
+     *
+     * @throws GuacamoleException
+     *     When an error occurs parsing the URI.
+     */
+    public static GuacamoleConfiguration getConfiguration(String uri)
+            throws GuacamoleException {
+
+        URI qcUri;
+        try {
+            qcUri = new URI(uri);
+        }
+        catch (URISyntaxException e) {
+            throw new GuacamoleClientException("Invalid URI Syntax", e);
+        }
+        String protocol = qcUri.getScheme();
+        String host = qcUri.getHost();
+        int port = qcUri.getPort();
+        String userInfo = qcUri.getUserInfo();
+        String query = qcUri.getQuery();
+        String username = null;
+        String password = null;
+        List<String> paramList = null;
+
+        if (protocol == null || protocol.isEmpty())
+            protocol = DEFAULT_URI_PROTOCOL;
+
+        if (host == null || host.isEmpty())
+            host = DEFAULT_URI_HOST;
+
+        if (query != null && !query.isEmpty())
+            paramList = Arrays.asList(query.split("&"));
+
+        if (userInfo != null && !userInfo.isEmpty()) {
+
+            Matcher userinfoMatcher = userinfoPattern.matcher(userInfo);
+            if (userinfoMatcher.matches()) {
+                username = userinfoMatcher.group(USERNAME_GROUP);
+                password = userinfoMatcher.group(PASSWORD_GROUP);
+            }
+
+        }
+
+        GuacamoleConfiguration qcConfig = new GuacamoleConfiguration();
+        qcConfig.setProtocol(protocol);
+        qcConfig.setParameter("hostname",host);
+
+        if (port > 0)
+            qcConfig.setParameter("port", Integer.toString(port));
+
+        if (username != null && !username.isEmpty())
+            qcConfig.setParameter("username", username);
+
+        if (password != null && !password.isEmpty())
+            qcConfig.setParameter("password", password);
+
+        if (paramList != null) {
+            for (String parameter : paramList) {
```

Comment:
It might make sense and would aid readability here if the parsing of the query were delegated to another method. It also seems like the kind of code that has lots of edge cases that would benefit greatly from some unit tests.

Reply:
I can definitely break out the query parsing into its own method.

Never done unit tests before...another new experience for me :-).  I know there are a couple of examples in the rest of the Guacamole code that I can probably use as a starting place.

Reply:
Okay, basic unit testing implemented - that wasn't too bad!

Let me know if those tests look okay...